cmd/kubeapps-apis/plugins/resources/v1alpha1/server.go

```diff
@@ -13,15 +13,144 @@ limitations under the License.
 }
 
 func NewServer(configGetter core.KubernetesConfigGetter) *Server {
-	return &Server{}
+	return &Server{
+		clientGetter: func(ctx context.Context, cluster string) (kubernetes.Interface, dynamic.Interface, error) {
+			if configGetter == nil {
+				return nil, nil, status.Errorf(codes.Internal, "configGetter arg required")
+			}
+			config, err := configGetter(ctx, cluster)
+			if err != nil {
+				return nil, nil, status.Errorf(codes.FailedPrecondition, "unable to get config : %v", err.Error())
+			}
+			dynamicClient, err := dynamic.NewForConfig(config)
+			if err != nil {
+				return nil, nil, status.Errorf(codes.FailedPrecondition, "unable to get dynamic client : %s", err.Error())
+			}
+			typedClient, err := kubernetes.NewForConfig(config)
+			if err != nil {
+				return nil, nil, status.Errorf(codes.FailedPrecondition, "unable to get typed client: %s", err.Error())
+			}
+			return typedClient, dynamicClient, nil
+		},
+		corePackagesClientGetter: func() (pkgsGRPCv1alpha1.PackagesServiceClient, error) {
+			port := os.Getenv("PORT")
+			conn, err := grpc.Dial("localhost:"+port, grpc.WithInsecure())
+			if err != nil {
+				return nil, status.Errorf(codes.Internal, "unable to dial to localhost grpc service: %s", err.Error())
+			}
+			return pkgsGRPCv1alpha1.NewPackagesServiceClient(conn), nil
+		},
+	}
+}
+
+// GetResources returns the resources for an installed package.
+func (s *Server) GetResources(r *v1alpha1.GetResourcesRequest, stream v1alpha1.ResourcesService_GetResourcesServer) error {
+	namespace := r.GetInstalledPackageRef().GetContext().GetNamespace()
+	cluster := r.GetInstalledPackageRef().GetContext().GetCluster()
+	log.Infof("+resources GetResources (cluster: %q, namespace=%q)", cluster, namespace)
+
+	ctx, err := copyAuthorizationMetadataForOutgoing(stream.Context())
+	if err != nil {
+		return err
+	}
+
+	// First we grab the resource references for the specified installed package.
+	coreClient, err := s.corePackagesClientGetter()
+	if err != nil {
+		return err
+	}
+	refsResponse, err := coreClient.GetInstalledPackageResourceRefs(ctx, &pkgsGRPCv1alpha1.GetInstalledPackageResourceRefsRequest{
+		InstalledPackageRef: r.InstalledPackageRef,
+	})
+	if err != nil {
+		return err
+	}
+
+	// Then look up each referenced resource and send it down the stream.
+	// TODO(minelson): Filter to the resources specified in the request, and 400
+	// if they don't exist for the package.
+	_, dynamicClient, err := s.clientGetter(stream.Context(), cluster)
+	if err != nil {
+		return err
+	}
+	for _, ref := range refsResponse.GetResourceRefs() {
+		groupVersion, err := schema.ParseGroupVersion(ref.ApiVersion)
+		if err != nil {
+			return status.Errorf(codes.Internal, "unable to parse group version from %q: %s", ref.ApiVersion, err.Error())
+		}
+		gvk := groupVersion.WithKind(ref.Kind)
+		// TODO(minelson): Find alternative to UnsafeGuessKindToResource.
+		gvr, _ := meta.UnsafeGuessKindToResource(gvk)
```

Comment:
Haha, the implementation is interesting (https://github.com/kubernetes/apimachinery/blob/master/pkg/api/meta/restmapper.go#L126). Plurals in k8s have always been a pain

Reply:
Yup.